dwave/cloud/solver.py

```diff
@@ -617,6 +619,98 @@ def upload_dqm(self, dqm):
         return self.upload_problem(dqm)
 
 
+class CQMSolver(BaseUnstructuredSolver):
+    """Class for D-Wave unstructured constrained quadratic model solvers.
+
+    This class provides a :term:`CQM` sampling method and encapsulates the
+    solver description returned from the D-Wave cloud API.
+
+    Args:
+        client (:class:`~dwave.cloud.client.Client`):
+            Client that manages access to this solver.
+
+        data (`dict`):
+            Data from the server describing this solver.
+
+    Note:
+        Events are not yet dispatched from unstructured solvers.
+    """
+
+    _handled_problem_types = {"cqm"}
+    _handled_encoding_formats = {"bq"}
+
+    def _encode_problem_for_upload(self, cqm):
+        try:
+            data = cqm.to_file()
+        except Exception as e:
+            logger.debug("CQM conversion to file failed with %r, "
+                         "assuming data already encoded.", e)
+            # assume `cqm` given as file, ready for upload
+            data = cqm
+
+        logger.debug("Problem encoded for upload: %r", data)
+        return data
+
+    def sample_bqm(self, bqm, label=None, **params):
+        """Use for testing."""
+        try:
+            import dimod
+        except ImportError: # pragma: no cover
+            raise RuntimeError("Can't sample from 'cqm' without dimod. "
+                               "Re-install the library with 'cqm' support.")
+
+        # to sample BQM problems, we need to convert them to CQM
+        cqm = dimod.CQM.from_bqm(bqm)
+
+        # TODO: convert sampleset back
```

Comment:
in the BQM case there is no conversion back from sample set. For DQM you do need to do some reinterpretation.

Reply:
Hmm, interesting question. IMO since at the moment we don't enforce the feasibility fields by subclass, having "extra" data vectors is totally fine. In the future we will probably have a `ConstrainedSampleSet` or something like that at which point we'd want to do some conversion.